aya/src/maps/xdp/dev_map_hash.rs

```diff
@@ -0,0 +1,121 @@
+//! An hashmap of network devices.
+
+use std::os::fd::AsRawFd;
+
+use aya_obj::generated::{bpf_devmap_val, bpf_devmap_val__bindgen_ty_1};
+
+use crate::{
+    maps::{check_kv_size, hash_map, IterableMap, MapData, MapError, MapIter, MapKeys},
+    sys::bpf_map_lookup_elem,
+};
+
+use super::dev_map::DevMapValue;
+
+/// An hashmap of network devices.
+///
+/// XDP programs can use this map to redirect to other network
+/// devices.
+///
+/// # Minimum kernel version
+///
+/// The minimum kernel version required to use this feature is 5.4.
+///
+/// # Examples
+/// ```no_run
+/// # let mut bpf = aya::Bpf::load(&[])?;
+/// use aya::maps::xdp::DevMapHash;
+///
+/// let mut devmap = DevMapHash::try_from(bpf.map_mut("IFACES").unwrap())?;
+/// let flags = 0;
+/// let ifindex = 32u32;
+/// devmap.insert(ifindex, ifindex, None::<i32>, flags);
```

Comment:
Does this redirect to itself?

Reply:
No, it just makes it so `key == value`.

A common pattern to ease debugging and programming is to use the same key as the actual interface index. For example, you'd have two maps: a `ROUTES` map (e.g. `LpmTrie`) that resolves to a destination and egress interface, and a `DEVS` map (`DevMap[Hash]`). The egress interface from the `ROUTES` map would point to an entry in the `DEVS` map. To avoid having to maintain a two-way mapping between interface and map index, it's quite common to use the ifindex as the map index. So if your interface has an ifindex of 32, then you put it at key 32, and use 32 as the egress interface in the `ROUTES` map. This basically removes an indirection level.

```rust
struct Dest {
    nexhop: u32,
    egress_ifindex: u32,
}
#[map] static ROUTES: LpmTrie<u32, Dest> = LpmTrie::with_max_entries(1000);
#[map] static DEVS: DevMapHash = LpmTrie::with_max_entries(8);

let Some(dest) = ROUTES.get((*ip).daddr) else { return XDP_DROP };
(*ip).daddr = dest.nexthop;
DEVS.redirect(dest.egress_ifindex, XDP_DROP);
```